src/algorithms/mul_redc.rs

```diff
@@ -1,70 +1,289 @@
+    debug_assert!(carry_outer <= 1);
+    reduce1_carry(result, modulus, carry_outer > 0)
+}
+
+#[inline]
+#[must_use]
+fn less_than<const N: usize>(lhs: [u64; N], rhs: [u64; N]) -> bool {
+    for (lhs, rhs) in zip(lhs.iter().rev(), rhs.iter().rev()) {
+        match lhs.cmp(rhs) {
+            Ordering::Less => return true,
+            Ordering::Greater => return false,
+            Ordering::Equal => {}
         }
-        debug_assert!(carry == 0 || temp[temp.len() - 1] == 1);
+    }
+    // lhs == rhs
+    false
+}
+
+#[inline]
+#[must_use]
+#[allow(clippy::needless_bitwise_bool)]
+fn reduce1_carry<const N: usize>(value: [u64; N], modulus: [u64; N], carry: bool) -> [u64; N] {
+    let (reduced, borrow) = sub(value, modulus);
+    // TODO: Ideally this turns into a cmov, which makes the whole mul_redc constant
+    // time.
+    if carry | !borrow {
+        reduced
+    } else {
+        value
+    }
+}
+
+#[inline]
+#[must_use]
+fn sub<const N: usize>(lhs: [u64; N], rhs: [u64; N]) -> ([u64; N], bool) {
+    let mut result = [0; N];
+    let mut borrow = false;
+    for (result, (lhs, rhs)) in zip(&mut result, zip(lhs, rhs)) {
+        let (value, next_borrow) = borrowing_sub(lhs, rhs, borrow);
+        *result = value;
+        borrow = next_borrow;
+    }
+    (result, borrow)
+}
+
+/// Compute `lhs * rhs + add + carry`.
+/// The output can not overflow for any input values.
+#[inline]
+#[must_use]
+#[allow(clippy::cast_possible_truncation)]
+const fn carrying_mul_add(lhs: u64, rhs: u64, add: u64, carry: u64) -> (u64, u64) {
+    let wide = (lhs as u128)
+        .wrapping_mul(rhs as u128)
+        .wrapping_add(add as u128)
+        .wrapping_add(carry as u128);
+    (wide as u64, (wide >> 64) as u64)
+}
+
+/// Compute `2 * lhs * rhs + add + carry_lo + 2^64 * carry_hi`.
+/// The output can not overflow for any input values.
+#[inline]
+#[must_use]
+#[allow(clippy::cast_possible_truncation)]
+const fn carrying_double_mul_add(
+    lhs: u64,
+    rhs: u64,
+    add: u64,
+    carry_lo: u64,
+    carry_hi: bool,
+) -> (u64, u64, bool) {
+    let wide = (lhs as u128).wrapping_mul(rhs as u128);
+    let (wide, carry_1) = wide.overflowing_add(wide);
+    let carries = (add as u128)
+        .wrapping_add(carry_lo as u128)
+        .wrapping_add((carry_hi as u128) << 64);
+    let (wide, carry_2) = wide.overflowing_add(carries);
+    (wide as u64, (wide >> 64) as u64, carry_1 | carry_2)
+}
+
+// Helper while [Rust#85532](https://github.com/rust-lang/rust/issues/85532) stabilizes.
+#[inline]
+#[must_use]
+const fn carrying_add(lhs: u64, rhs: u64, carry: bool) -> (u64, bool) {
+    let (result, carry_1) = lhs.overflowing_add(rhs);
+    let (result, carry_2) = result.overflowing_add(carry as u64);
+    (result, carry_1 | carry_2)
+}
+
+// Helper while [Rust#85532](https://github.com/rust-lang/rust/issues/85532) stabilizes.
+#[inline]
+#[must_use]
+const fn borrowing_sub(lhs: u64, rhs: u64, borrow: bool) -> (u64, bool) {
```

Comment:
We define these exact functions in add.rs https://github.com/recmo/uint/blob/b91e6f710e42b893e695cae43ffab3d65740d881/src/add.rs#L61
We should probably move the definition to `algorithms` maybe?

Reply:
I think my idea was to keep the `mul_redc` file self-contained (I've been copying it between projects), but there's no reason to do this here. I'll update.